Sources/MapboxCoreNavigation/MapboxRoutingProvider.swift

```diff
@@ -364,8 +388,19 @@ public class MapboxRoutingProvider: RoutingProvider {
                     DispatchQueue.global().async {
                         do {
                             let routeResponse = try indexedRouteResponse.routeResponse.copy(with: routeOptions)
-                            routeResponse.routes?[indexedRouteResponse.routeIndex].refreshLegAttributes(from: routeRefreshResponse.route)
-                            routeResponse.routes?[indexedRouteResponse.routeIndex].refreshLegIncidents(from: routeRefreshResponse.route)
+                            let route = routeResponse.routes?[indexedRouteResponse.routeIndex]
+                            if let currentLegShapeIndex = currentLegShapeIndex {
+                                route?.refreshLegAttributes(from: routeRefreshResponse.route,
```

Comment:
As of mapbox/mapbox-directions-swift#733, [this method](https://github.com/mapbox/mapbox-directions-swift/blob/55c2672721e1443d5761c921ee61d48fd8852ccc/Sources/MapboxDirections/RouteRefreshResponse.swift#L124) accounts for the offset from the beginning of the route up to the current shape index; however, it fails to clear attributes along the earlier portion of the route. If the route line isn’t tracking the user’s location (in other words, if the route line doesn’t “vanish”), then the same traffic jam could appear multiple times along the route line as it and the user move along the route.

Moreover, the API counterintuitively reacts to this shape index by truncating the _end_ of the route as well. I think we should clear any stale congestion attributes from that latter portion of the route, past the end of this refresh, or else a stale traffic jam could unduly influence proactive rerouting. Distance and speed limit attributes won’t go stale, so we should keep those around. We would probably need to keep stale expected travel time attributes around to avoid skewing the ETA or fraction traveled.

/ref https://github.com/mapbox/mapbox-navigation-ios/pull/4111#discussion_r958792536

Reply:
1. up to the current shape index - clear all attributes and incidents;
2. from the current geometry index up to the (current geometry index + length of updated shape) - works as expected;
3. from the last updated index to the rest of the shape - clear only congestion levels.

@1ec5 is that your suggestion? Please elaborate on the case of refreshing the route from a specific geometry index while not tracking the user location.

I'd like to make this uniform across platforms, @mapbox/navigation-android how do you handle this?

Reply:
Are you asking about truncated end? For the end we use old annotations (all of them: distance, duration, congestion, etc.) but clear incidents and closures. 
Why do we not clear congestion? Not sure, I think it is just a common approach for annotations. We can discuss that. But as far as I know, by default unknown congestion has the same colour as normal so it could confuse a driver that after refresh red lines at the end of the route suddenly turned into blue ones. They may consider it a natural traffic conditions change while it's an artificial clearance  of traffic data.

Reply:
Thanks, @dzinad. And what about attributes from the start of the route to the current geometry index? Do you clear them or do you leave the stale data?

Reply:
The same: we leave the stale data for annotations and clear incidents and closures. But here it's necessary for annotations because they have to have specific indices that are aligned with geometry. You can't just remove them. While incidents and closures have these indices specified in themselves.

Reply:
Then to have a consistent implementation across platforms we need to clear incidents before and after the refresh range. Attributes are already in-sync.
Thinking about clearing attributes after refresh (meaning setting them to unknown/empty values) - later segments may suddenly stop being taken into account after refresh, so some metrics like ETA could jump in up or down only to jump back again when we refresh reaches these segments again.

JFYI this thread doesn't affect this particular PR and discusses already merged logic in mapbox-directions-swift.

Reply:
The parity in incidents/attributes handling on refresh is ensured in https://github.com/mapbox/mapbox-directions-swift/pull/752.